Implement python unit test cases (#267)
pez-globo/pufferfish-vent-software: backend/tests/unit/protocols/backend.txt

When: a new value is set for the frontend field of the object
Then: has_data function returns true

Scenario: The backend's output event correctly reports whether it has data
Given: A OutputEvent object constructed with a value for the mcu, frontend and file parameter
When: a new value is set for the file field of the object
Then: has_data function returns true

Scenario: The input event sent by backend correctly reports whether it has data
Given: a announcement object constructed with default parameters
When: the object is not modified
Then: has_data function returns false

Scenario: The input event sent by backend correctly reports whether it has data
Given: a announcement object constructed a value for the message field
When: a new value is set for the message field of the object
Then: has_data function returns true

#ReceiveFilter

Scenario: The receive filter's output correctly reports whether it has data
Given: a ReceiveFilter object which has not yet received any events
When: a receive event with no data is passed as input to the receive filter
Then: the output of the receive filter is None

Scenario: The receive filter's output correctly reports whether it has data

> Given: A ReceiveEvent object constructed with values for the time and mcu parameter

Reviewer: On these tests, we need to be given a receive filter object. If you are constructing a different ReceiveEvent object for each test, it might make more sense to define it in the When. So, for example:

Given: A ReceiveFilter object constructed with ___ and which has an empty internal input events buffer
(Then: The output method of the ReceiveFilter object returns None each time it is called)
When: A ReceiveEvent object with values for ___ is given as input to the ReceiveFilter object
Then: The output method of the ReceiveFilter object returns None each time it is called

> Scenario: Backend receive event behaves properly

Reviewer: ```suggestion
Scenario: The backend receiver's input event behaves properly
```
Rationale: this wording for the scenario is more specific, and it is a bit more descriptive about the class we're testing, by associating it with the backend's receiver (which is implemented by the `ReceiveFilter` class).

Author: Changed as per suggestion

Reviewer: I still see this wording in the latest commit - have you pushed all your changes?

Reviewer: I've seen the wording you used for scenarios in your tests of event classes in `tests/unit/protocols/server.txt`, and I like it. Maybe we should change this to "Scenario: The backend receiver's input event correctly reports whether it has data" (and apply this pattern to all other relevant tests in this PR, too).

Author: i think the comments were written for the particular commit-id, therefore it is not visible with new commits? comments written in the main pr page seem to be visble for all new commits when pressed on "view changes"

Reviewer: For whatever reason, it had still shown up under the "Files Changed" tab which usually only shows comments which still apply to the latest changes. But it's gone away, so it's not an issue anymore.

Given: A ReceiveEvent object

> When: ReceiveEvent class is initialised without data

Reviewer: ```suggestion
When: the object is initialised without data
```

or

```suggestion
When: no data is given as input
```

Rationale: the things we initialize are objects, not classes

Author: Changed as per suggestion

> Then: has_data function returns false   

Reviewer: Based on what we discussed today with the `CRCElement` class in the C++ code, I wonder if it might be more consistent with other tests to structure these tests differently, so that the Given corresponds to our instantiation of the object. For example:

Given: A ReceiveEvent object constructed with default parameters
When: the object is not modified
Then: the has_data method returns false

Given: A ReceiveEvent object constructed with default parameters
When: a new value is set for the time field of the object
Then: the has_data method returns true

Given: A ReceiveEvent object constructed with a value for the time parameter
When: the object is not modified
Then: the has_data method returns true

Given: A ReceiveEvent object constructed with a value for the time parameter
When: the time field of the object is set to None
Then: the has_data method returns false

and so on.

Author: I have structured the descriptions as per the suggestions here : 

https://github.com/pez-globo/pufferfish-software/blob/feature/backend-test-protocols2/backend/tests/unit/protocols/backend.txt
please review.

Reviewer: Looks good.

Scenario: Backend receive event behaves properly
Given: A ReceiveEvent object

> When: time is passed as input

Reviewer: ```suggestion
When: time is given as input
```
Rationale: I'm being pedantic, but in these ReceiveEvent object tests I would suggest using `given` rather than `passed` - since "passing" evokes passing values into a function (e.g. "pass by value" and "pass by reference" in C++ terminology), and here we're testing a method which checks public data members.

Author: Changed as per suggestion

Then: has_data function returns true

Scenario: Backend receive event behaves properly
Given: A ReceiveEvent object
When: mcu event is passed as input
Then: has_data function returns true

Scenario: Backend receive event behaves properly
Given: A ReceiveEvent object
When: frontend data is passed as input
Then: has_data function returns true

> Scenario: Backend receive event behaves properly

Reviewer: ```suggestion
Scenario: The backend's output event behaves properly
```
Rationale: this event is used for the backend receiver's output, the backend sender's output, and the backend sender's input.

Author: Changed as per suggestion

Given: A OutputEvent object
When: ouptut event has no data
Then: has_data function returns false

Scenario: Backend receive event behaves properly
Given: A OutputEvent object
When: output event has data
Then: has_data function returns true

> Scenario: Backend send event behaves properly

Reviewer: ```suggestion
Scenario: The backend's output event behaves properly
```

Author: Changed as per suggestion

Given: a announcement object
When: backend recieves announcement strings as input
Then: has_data function returns true

#ReceiveFilter

> 

Reviewer: We will also need to write a test for what we expect to happen when we pass `None` into the `input` method. And we will need to write a test for what we expect to be returned by `output` when we have not passed any data into the `input` method, and when we called `input` a few times but we've also already called `output` more times.

Author: Will push new tests in a different pull request

Scenario: Receive Filter behaves properly

> Given: receive event with no data

Reviewer: ```suggestion
Given: a `ReceiveFilter` object which has not yet received any events
```
Of course, we will also need to write tests for `ReceiveFilter` objects which have already received various sequences of events.

Author: Changed as per suggestion, new tests will push in a different pull request

> Given: a `ReceiveFilter` object which has not yet received any events

Reviewer: ```suggestion
Given: a ReceiveFilter object which has not yet received any events
```
Oops, I had accidentally used Markdown when I made the code change suggestion for this line.

Author: Updated. check here https://github.com/pez-globo/pufferfish-software/blob/feature/backend-test-protocols2/backend/tests/unit/protocols/backend.txt

Reviewer: Looks good.

> When: empty receive event is passed as input to receive filter

Reviewer: ```suggestion
When: a receive event with no data is passed as input to the receive filter
```

Author: Changed as per suggestion

> Then: output of receive filter None

Reviewer: ```suggestion
Then: the output of the receive filter is None
```

Author: Changed as per suggestion

> When: a receive event with no data is passed as input to the receive filter

Reviewer: We could also, for example, add a 'Then: the internal buffer is empty', which would correspond to checking the state of the `_buffer` "private" member - but that test may not be so useful here. In other cases where there's complicated logic (particularly in the `output` methods of the backend and server protocols, as well as in the `input` and `output` methods of application protocols such as for list synchronization), we need to write some tests on the state of private members before and after method calls, to ensure that assumptions, preconditions, and invariants are maintained by the methods. Anyways, you can keep those for a future PR.

Author: Noted, I intentionally avoided checking private members at most places because of lint checks, I'll keep in mind when writing future tests. thanks

Reviewer: Got it. Yes, for tests it's always fine to check the values of private members and add a manual suppression for pylint

Scenario: Receive Filter behaves properly

> Given: mcu receive event

Reviewer: You will also need to update the Givens and Whens for this scenario and the next six scenarios, in a similar way to how I had previously suggested changes to the scenario above.

Author: Updated in the latest commit for the file.

Reviewer: I'll add a comment on the changed version of the descriptions.

When: mcu receive event is passed as input to receive filter

> Then: output of receive filter is not None

Reviewer: We will also need to write tests which make more specific assertions about what the contents of the output of the receive filter should look like, under different conditions. This will probably require writing multiple new tests.

Author: Will push new tests in a different pull request

Scenario: Receive Filter behaves properly
Given: frontend receive event
When: frontend receive event is passed as input to receive filter
Then: output of receive filter is not None

Scenario: Receive Filter behaves properly
Given: file object receive event
When: file receive event is passed as input to receive filter
Then: output of receive filter is not None

#SendFilter

Scenario : Send Filter behaves properly
Given: An Event of type SendEvent with no data
When: empty send event is passed as input to send filter
Then: output of send filter is None

Scenario : Send Filter behaves properly
Given: An event of type announcement
When: announcement event is passed as input to send filter
Then: output of send filter is not None

Scenario : Send Filter behaves properly
Given: An Event of type SendEvent with data
When: Given event is passed as input to send filter
Then: output of send filter is not None

#Protocols

> 

Reviewer: We will also need to write tests where `Then` is `the output of the function is None`.

Author: Will push new tests in a different pull request

Scenario: get_mcu_send function behaves properly

> Given: A OutputEvent from mcu

Reviewer: ```suggestion
Given: The get_mcu_send function
```
I think the `Given`s should be the functions.

Author: Resolved

> When: given event is passed to get_mcu_send function

Reviewer: ```suggestion
When: an OutputEvent object containing data in the mcu_send field is passed to the function
```
Maybe this is too specific to the names of the data members of the `OutputEvent` class, but we need to be at least somewhat specific.

Author: updated as per suggestion
